### ISSUE TITLE
Can't reload in 1.18.1

### ISSUE THREAD

**Issue by SRR8** [MrCrayfish/MrCrayfishGunMod#223]
Installed the mod, could fire and reload. Came back the next day with no mod changes, and now no-one can reload.

**Comment by MrCrayfish** [comment]
You probably installed a mod that overrides the R button.

**Comment by SRR8** [comment]
I thought so too, so I set the keybind to other keybinds with no conflicts and still no luck.
Even so it worked when first installed but when we loaded up the server the next day it won't reload at all.

**Comment by SRR8** [comment]
I've fixed it. I removed it from the server, restarted the server without the mod, then restarted the server with the mod back on. No idea why this worked but it did.

**Comment by MrCrayfish** [comment]
Let me know if it happens again. Did you restart the server between the off time when the mod was installed?

**Comment by Diamantino-Op** [comment]
I'm having this problem too, i even tried recompiling mod from src too with no luck.

@MrCrayfish i may have found out after some debugging what is the problem

The code: if(ModSyncedDataKeys.RELOADING.getValue(player)) of class ReloadTracker can't get the value of the RELOADING Key

Edit:
found out more, this time is not in the Gun Mod but in the mod used to get and set the values of RELOADING (FrameWork), in the SyncedEntityData, in the methods get and set, the DataHolder holder = this.getDataHolder(entity); is always null.

**Comment by MrCrayfish** [comment]
Can you tell me the steps to reproduce?

**Comment by Diamantino-Op** [comment]
I just loaded up a test world and tried reloading

Btw, unloading works

For the denug, just put some System.out.println in the class

**Comment by MrCrayfish** [comment]
You said `DataHolder holder = this.getDataHolder(entity);` is returning null but I don't get that result, even after reloading a world.

**Comment by Diamantino-Op** [comment]
idk, i get null if i set a println in that

**Comment by MrCrayfish** [comment]
Well not all entities have a data holder, only ones that have a Synced Key registered to them.

**Comment by Diamantino-Op** [comment]
I modified like this:

public <E extends Entity, T> void set(E entity, SyncedDataKey<?, ?> key, T value)
    {
        if(!this.registeredDataKeys.contains(key))
        {
            throw new IllegalArgumentException(String.format("The synced data key %s for %s is not registered!", key.id(), key.classKey().id()));
        }
        DataHolder holder = this.getDataHolder(entity);
        System.out.println(CAPABILITY);
        if(holder != null && holder.set(entity, key, value))
        {
            if(!entity.level.isClientSide())
            {
                System.out.println("Set Data:" + key + " On: " + entity);
                this.dirty = true;
                this.dirtyEntities.add(entity);
            }
            else
                System.out.println("Entity is ClientSide");
        }
        else
            System.out.println("holder is null");
    }

**Comment by Diamantino-Op** [comment]
and it only give result when i press R to reload

**Comment by MrCrayfish** [comment]
Well `if(holder != null && holder.set(entity, key, value))` does not mean holder is null. It's possible for `set` to return false.

**Comment by Diamantino-Op** [comment]
then, what is causing the not reload bug?

**Comment by MrCrayfish** [comment]
Again, with no steps given to reproduce the issue, it's a lot harder to find the issue.

**Comment by Diamantino-Op** [comment]
can you reload if you run from src?

**Comment by MrCrayfish** [comment]
Yes, I developed the mod.

**Comment by Diamantino-Op** [comment]
ok, i tried some more debug, and it arrives here:

@Override
    public void handle(MessageReload message, Supplier<NetworkEvent.Context> supplier)
    {
        supplier.get().enqueueWork(() ->
        {
            ServerPlayer player = supplier.get().getSender();
            if(player != null && !player.isSpectator())
            {
                ModSyncedDataKeys.RELOADING.setValue(player, message.reload); // This has to be set in order to verify the packet is sent if the event is cancelled

                System.out.println("Player Name: " + player + " Reload Bool: " + message.reload);

                if(!message.reload)
                    return;

                ItemStack gun = player.getMainHandItem();

                System.out.println("Gun: " + gun.getDisplayName());

                if(MinecraftForge.EVENT_BUS.post(new GunReloadEvent.Pre(player, gun))) <-------- HERE
                {
                    System.out.println("GunReload Pre");
                    ModSyncedDataKeys.RELOADING.setValue(player, false);
                    return;
                }
                MinecraftForge.EVENT_BUS.post(new GunReloadEvent.Post(player, gun));
            }
        });
        supplier.get().setPacketHandled(true);
    }

**Comment by Diamantino-Op** [comment]
is it normal?

**Comment by MrCrayfish** [comment]
Well if it's cancelled by the event, it's probably another mod cancelling the reloading.

**Comment by Diamantino-Op** [comment]
what other mod?

**Comment by Diamantino-Op** [comment]
i have only the three required

**Comment by Diamantino-Op** [comment]
oh wait, maybe JEI?

**Comment by MrCrayfish** [comment]
Okay, it's probably the event since it's only allowing the gun to reload if the event is cancelled.

**Comment by Diamantino-Op** [comment]
ah

**Comment by MrCrayfish** [comment]
Actually it's not the event. I didn't read the code completely.

**Comment by Diamantino-Op** [comment]
then what is that?, bc it stops there the content of the if is never executed

**Comment by MrCrayfish** [comment]
Read the code

**Comment by Diamantino-Op** [comment]
ah yes, it stops reloading, my bad

**Comment by Diamantino-Op** [comment]
Btw, now i can confirm that my holder is null when i try to reload

**Comment by Diamantino-Op** [comment]
I added this code:

if (holder != null)
            System.out.println("Key: " + key + " Value: " + holder.set(entity, key, value) + " For: " + entity + " Value2: " + value);
        else
            System.out.println("Key: " + key + " Value: " + "Holder Null" + " For: " + entity + " Value2: " + value);

and it gave me "Holder Null" with the key cgm:reloading.

**Comment by MrCrayfish** [comment]
Again Holder is not null for me.

**Comment by Diamantino-Op** [comment]
idk then

**Comment by MrCrayfish** [comment]
It's possible that I misconfigured the entity capability. Since you're able to replicate the issue.

In Framework, go to `SyncedEntityData` then `Provider` and replace `getCapability` with this below code.
```java
@Nonnull
@Override
public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> cap, @Nullable Direction side)
{
    return cap == CAPABILITY ? this.optional.cast() : LazyOptional.empty();
}
```
then test it with the Gun Mod.

**Comment by Diamantino-Op** [comment]
gimme a sec

**Comment by Diamantino-Op** [comment]
nope

**Comment by Diamantino-Op** [comment]
@MrCrayfish i fixed my issue, turns out that during debug i accidentally deleted @SubscribeEvent from the attachCapability function in the SyncedEntityData...

**Comment by Diamantino-Op** [comment]
now it all works, and that fix should fix the previous bug

**Comment by Nightmarionne-1** [comment]
idk what to do i have the same prob but i have no idea what to do ;w;

**Comment by MrCrayfish** [comment]
Fixed in latest versions of 1.18.2 and above
